File: packages/dart/lib/src/catalogs/sat/cfdi/tipo_comprobante.dart
/// Catálogo de Tipos de Comprobante (SAT)
///
/// Catálogo c_TipoDeComprobante del SAT para CFDI 4.0
library;

import 'package:catalogmx/src/catalogs/base_catalog.dart';

/// Tipo de Comprobante Catalog
class TipoComprobanteCatalog {
  static List<Map<String, dynamic>>? _data;
  static Map<String, Map<String, dynamic>>? _byClave;

  static void _loadData() {
    if (_data != null) return;

    final jsonData = BaseCatalog.loadJsonDataSync(
      'sat/cfdi_4.0/tipo_comprobante.json',
    );

    // Handle both list and dict formats
    if (jsonData.isNotEmpty && jsonData.first.containsKey('tipos')) {
      _data = (jsonData.first['tipos'] as List)
          .map((e) => e as Map<String, dynamic>)
          .toList();
    } else {
      _data = jsonData;
    }

    // Build index
    _byClave = {for (var item in _data!) item['clave'] as String: item};
  }

  /// Obtiene todos los tipos de comprobante
  static List<Map<String, dynamic>> getAll() {
    _loadData();
    return List.from(_data!);
  }

  /// Obtiene tipo de comprobante por clave
  static Map<String, dynamic>? getByClave(String clave) {
    _loadData();
    return _byClave![clave];
  }

  /// Valida si una clave de tipo de comprobante existe
  static bool isValid(String clave) {
    return getByClave(clave) != null;
  }
}

File: packages/python/catalogmx/catalogs/sat/cfdi_4/tipo_comprobante.py
"""Catálogo c_TipoComprobante"""

from catalogmx.catalogs.sat.cfdi_4._resolver_views import code_description_rows


class TipoComprobanteCatalog:
    """Catálogo de Tipos de Comprobante del SAT (c_TipoComprobante)"""

    _data: list[dict] | None = None
    _by_code: dict[str, dict] | None = None

    @classmethod
    def _load_data(cls) -> None:
        """Carga lazy desde el artefacto canónico CFDI 4.0."""
        if cls._data is None:
            cls._data = code_description_rows("cfdi_40_tipos_comprobantes")
            cls._by_code = {item["code"]: item for item in cls._data}

    @classmethod
    def get_tipo(cls, code: str) -> dict | None:
        """Obtiene un tipo de comprobante por su código"""
        cls._load_data()
        return cls._by_code.get(code.upper())

    @classmethod
    def is_valid(cls, code: str) -> bool:
        """Valida si un código de tipo de comprobante es válido"""
        return cls.get_tipo(code) is not None

    @classmethod
    def get_all(cls) -> list[dict]:
        """Obtiene todos los tipos de comprobante"""
        cls._load_data()
        return cls._data.copy()
